### PR TITLE
chore: documentation for curve module

### PR DESCRIPTION
**Pull request by Ad96el**
adds some docs for the curve module.

### CODE REVIEW

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs [hunk 1]
```diff
@@ -1,3 +1,28 @@
...
+/// - `s` is the supply of assets,
+/// - `m` is the liquidity parameter of the LMSR model,
+/// - `s_i` is the current supply of the assets.
+///
+/// `C(s)` represents the accumulated cost of purchasing/selling assets up to the current supply `s`.
+/// To calculate the incremental cost of purchasing the assets, use the formula:
+/// `C(s) - C(s*)`, where `s*` is the supply of assets in the market before the purchase.
+///
+///
+/// Components
```
Comment (ntn-x2) [inline pallets/pallet-bonded-coins/src/curves/lmsr.rs:18]: I don't think it's necessary to list these components here. They will just be exported in the doc of they are public. This is dangerous as it risks (almost certainly) to get out-of-date if we change anything in this module. I would simply remove this list here.

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs [hunk 2]
```diff
@@ -1,3 +1,28 @@
...
+/// - `LMSRParametersInput`: A struct representing the input parameters for the LMSR model.
+/// - `LMSRParameters`: A struct representing the parameters for the LMSR model. Used to perform calculations and stored in storage.
+/// - `TryFrom<LMSRParametersInput<I>> for LMSRParameters<C>`: An implementation to convert input parameters to the correct fixed-point type.
+/// - `BondingFunction<Parameter> for LMSRParameters<Parameter>`: An implementation of the bonding function to calculate costs.
+///
+/// Optimization
+/// For numerical stability, the LMSR bonding curve employs the log-sum-exp trick to calculate the cost.
```
Comment (ntn-x2) [inline pallets/pallet-bonded-coins/src/curves/lmsr.rs:25]: Link to the trick?

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs [hunk 3]
```diff
@@ -11,16 +36,30 @@ use substrate_fixed::{
 use super::BondingFunction;
 use crate::{PassiveSupply, Precision, LOG_TARGET};
 
+/// A struct representing the input parameters for the LMSR model. This struct is used to convert
+/// the input parameters to the correct fixed-point type.
+///
+/// # Fields
```
Comment (ntn-x2) [inline pallets/pallet-bonded-coins/src/curves/lmsr.rs:42]: Why not documenting the field itself, and rely on cargo capabilities to generate the docs?

File: pallets/pallet-bonded-coins/src/curves/mod.rs [hunk 4]
```diff
@@ -87,6 +112,14 @@ where
 	}
 }
 
+/// Trait defining the bonding function for a curve.
+/// The bonding function is used to calculate the cost of purchasing or selling assets using the curve.
+/// The trait is implemented for each curve type.
+///
+/// Variables:
```
Comment (ntn-x2) [inline pallets/pallet-bonded-coins/src/curves/mod.rs:119]: These are called arguments.

**Review by ntn-x2 — COMMENTED** [review]
Comments left on the lmsr module apply to the other modules as well.

File: pallets/pallet-bonded-coins/src/curves/mod.rs (hunk 4 above)
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/mod.rs]: Updated in: [2ae4290](https://github.com/KILTprotocol/kilt-node/pull/801/commits/2ae429067b29b8adcd5200855c7b3e4dcfa0abe2)

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs [hunk 5]
```diff
@@ -1,3 +1,28 @@
+/// Implementation of the Logarithmic Market Scoring Rule (LMSR) bonding curve.
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/lmsr.rs]: Updated in: [2ae4290](https://github.com/KILTprotocol/kilt-node/pull/801/commits/2ae429067b29b8adcd5200855c7b3e4dcfa0abe2)

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs [hunk 6]
```diff
@@ -1,3 +1,28 @@
...
+/// - `m` is the liquidity parameter of the LMSR model,
+/// - `s_i` is the current supply of the assets.
+///
+/// `C(s)` represents the accumulated cost of purchasing/selling assets up to the current supply `s`.
+/// To calculate the incremental cost of purchasing the assets, use the formula:
+/// `C(s) - C(s*)`, where `s*` is the supply of assets in the market before the purchase.
+///
+///
+/// Components
+/// - `LMSRParametersInput`: A struct representing the input parameters for the LMSR model.
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/lmsr.rs]: The section is dropped in: [2ae4290](https://github.com/KILTprotocol/kilt-node/pull/801/commits/2ae429067b29b8adcd5200855c7b3e4dcfa0abe2)

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs [hunk 7]
```diff
@@ -1,3 +1,28 @@
...
+/// - `LMSRParameters`: A struct representing the parameters for the LMSR model. Used to perform calculations and stored in storage.
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/lmsr.rs]: Section doped in: [2ae4290](https://github.com/KILTprotocol/kilt-node/pull/801/commits/2ae429067b29b8adcd5200855c7b3e4dcfa0abe2)

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs (hunk 1 above)
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/lmsr.rs]: Section droped in: [2ae4290](https://github.com/KILTprotocol/kilt-node/pull/801/commits/2ae429067b29b8adcd5200855c7b3e4dcfa0abe2)

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs (hunk 3 above)
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/lmsr.rs:35]: Good point. Wasn't thinking of it: [2ae4290](https://github.com/KILTprotocol/kilt-node/pull/801/commits/2ae429067b29b8adcd5200855c7b3e4dcfa0abe2)

File: pallets/pallet-bonded-coins/src/curves/mod.rs [hunk 8]
```diff
@@ -116,7 +111,7 @@ where
 /// The bonding function is used to calculate the cost of purchasing or selling assets using the curve.
 /// The trait is implemented for each curve type.
 ///
-/// Variables:
+/// Called arguments:
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-bonded-coins/src/curves/mod.rs:114]: In a function definition, they are called "arguments". In a function invocation, they are typically referred to as "parameters".

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs [hunk 9]
```diff
@@ -1,3 +1,22 @@
...
+/// The cost of purchasing an amount of assets from the market is determined using the liquidity parameter of the
+/// LMSR model and the current supply of assets in the market.
+/// The LMSR bonding curve is defined by the following equation:
+///
+/// C(s) = m * ln(Σ(e^(s_i/m))),
+///
+/// where:
+/// - `s` is the supply of assets,
+/// - `m` is the liquidity parameter of the LMSR model,
+/// - `s_i` is the current supply of the assets.
```
Comment (rflechtner) [inline pallets/pallet-bonded-coins/src/curves/lmsr.rs:11]: can you make the difference between s and s_i clearer? Is s the cumulated total supply of all assets, and s_i the supply of one asset?

File: pallets/pallet-bonded-coins/src/curves/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply pallets/pallet-bonded-coins/src/curves/mod.rs:114]: is it not the other way around?

[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/pallet-bonded-coins/src/curves/mod.rs:114]: Ops, you are right 🙈

File: pallets/pallet-bonded-coins/src/curves/polynomial.rs [hunk 10]
```diff
@@ -1,3 +1,24 @@
+/// Polynomial bonding curve implementation.
+///
+/// This module provides an implementation of a polynomial bonding curve. The current implementation supports a square bonding curve, where the integral is precomputed.
+/// The cost function for the polynomial bonding curve is defined as:
+/// C(s) = m * s^3 + n * s^2 + o * s,
+/// where:
+/// - `s` is the supply of assets,
+/// - `m` is the coefficient for the cubic part,
+/// - `n` is the coefficient for the quadratic part,
+/// - `o` is the coefficient for the linear part.
```
Comment (rflechtner) [inline pallets/pallet-bonded-coins/src/curves/polynomial.rs:10]: it probably helps to explain what m, n, and o stand for in the price function, and how to derive the parameters of the price function from the given values

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs [hunk 11]
```diff
@@ -36,7 +29,7 @@ use substrate_fixed::{
 use super::BondingFunction;
 use crate::{PassiveSupply, Precision, LOG_TARGET};
 
-/// A struct representing the input parameters for the LMSR model. This struct is used to convert
+/// A struct representing the unchecked input parameters for the LMSR model. This struct is used to convert
 /// the input parameters to the correct fixed-point type.
 ///
 /// # Fields
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/lmsr.rs]: Oh, sorry. I double checked, there should no `Fields` anymore: [8f8664a](https://github.com/KILTprotocol/kilt-node/pull/801/commits/8f8664a534ff38f5e7f491145f18f69f0eb1019f)

File: pallets/pallet-bonded-coins/src/curves/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/mod.rs]: Is parameter right now: [8f8664a](https://github.com/KILTprotocol/kilt-node/pull/801/commits/8f8664a534ff38f5e7f491145f18f69f0eb1019f)?

File: pallets/pallet-bonded-coins/src/curves/polynomial.rs [hunk 12]
```diff
@@ -1,3 +1,24 @@
+/// Polynomial bonding curve implementation.
+///
+/// This module provides an implementation of a polynomial bonding curve. The current implementation supports a square bonding curve, where the integral is precomputed.
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/polynomial.rs]: rephrased in: [8f8664a](https://github.com/KILTprotocol/kilt-node/pull/801/commits/8f8664a534ff38f5e7f491145f18f69f0eb1019f)

File: pallets/pallet-bonded-coins/src/curves/polynomial.rs (hunk 10 above)
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/polynomial.rs]: I added more explanation in: [8f8664a](https://github.com/KILTprotocol/kilt-node/pull/801/commits/8f8664a534ff38f5e7f491145f18f69f0eb1019f) 

Do you think it is clearer now?

File: pallets/pallet-bonded-coins/src/curves/square_root.rs [hunk 13]
```diff
@@ -1,3 +1,20 @@
+/// Square Root bonding curve implementation.
+///
+/// This module provides an implementation of a square root bonding curve. The current implementation supports a bonding curve where the integral is precomputed.
+/// The cost function for the square root bonding curve is defined as:
+/// C(s) = m * sqrt(s^3) + n * s,
+/// where:
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/square_root.rs]: Added more explanation in: [8f8664a](https://github.com/KILTprotocol/kilt-node/pull/801/commits/8f8664a534ff38f5e7f491145f18f69f0eb1019f)

File: pallets/pallet-bonded-coins/src/curves/lmsr.rs (hunk 9 above)
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/lmsr.rs]: There was a little mistake. :grimacing:  Fixed in: [8f8664a](https://github.com/KILTprotocol/kilt-node/pull/801/commits/8f8664a534ff38f5e7f491145f18f69f0eb1019f)

**Review by ntn-x2 — APPROVED** [review]

**Review by rflechtner — APPROVED** [review]

**Pull request merged**